feat(core): document attach partition when partition.attachable is a symlink (#1176)
questdb/questdb.io: docs/reference/sql/alter-table-attach-partition.md

description: ATTACH PARTITION SQL keyword reference documentation.
---

Restores one or more partitions to the table where they have been detached from by using the SQL [ALTER TABLE DETACH PARTITION](/docs/reference/sql/alter-table-detach-partition) statement. 

This feature is part of the manual S3/cold storage solution, allowing restoring data manually.

Before executing `ATTACH PARTITION`, the partition folders to be attached must be made available to QuestDB. There are two approaches: 
1. copy the partition folder into the table folder and rename it to `<partition_name>.attachable`, e.g given a partition named `2020.detached`, copy as `2020.attachable`. 
2. create a symlink of name `<partition_name>.attachable` from the table folder to the partition folder. The partition folder can be in a different volume. 

Partitions attached via the symlink approach are considered **read only** in these operations:

> - _insert_, _add index_/_drop index_, _add column_/_drop column_, _rename column_: the partition is copied to the table folder under a new transaction version, where the operation is executed. 

Author: Hi Amy, I have amended the text to include your comments, in particular:
- I have capitalised the keywords
- the list has been amended to reflect the current state of questdb/master. There is no copy involved, except in the case when an insert into the partition comes out of order (not worth including in the documentation page, too much detail for an unusual situation).

>   The copy operation will be costly if the partition volume has a slower disk. The original partition folder is not changed in any way.

Reviewer: This PR also seems to contradict the original [PR description](https://github.com/questdb/questdb/pull/2650): 
>The original copy of the partition attached via soft link will remain READ ONLY for detach/drop partition operations.
>The original copy of the partition attached via soft link WILL BE CHANGED by insert/update operations, as well as add/drop index, add/drop/rename column operations.

>   The copy operation will be costly if the partition volume has a slower disk. The original partition folder is not changed in any way.

Author: I have added an info sandwich in the document, to make users aware of disk speeds and their effect on SQL statements that hit partitions attached via symlink.

symlinks are a normal file system concept. They are a file which refers to another file. You have to imagine files as labelled boxes that contain data. A symlink is like a box, it has a label too, but it is empty, instead the bottom is broken and you magically enter the box the symlink is pointing to. So in effect, opening the symlink box is like opening the target box, except you can hit the symlink box with a hammer and the data within the target box will be ok. I have added a wikipedia link.

I have updated the text in general to reflect the current state of questdb/master, where attaching via symlink is not read-only yet.

> - _drop partition_/_detach partition_: the partition is unlinked, the symlink is removed but not the content of the partition folder. Detaching a partition that was attached via symlink does not 

Author: the detach page points right back to this page.

`2020.detached`, rename it as `2020.attachable` and move it to the correct table
folder.

### Symbolic links

[Symbolic links](https://en.wikipedia.org/wiki/Symbolic_link) are soft links
pointing to partition folders in cold storage. To make partition folders
available for attaching, set the target path to the name
`<partition_name>.attachable`.

As symbolic links are soft links, the folders can be in a different volume.

:::info

SQL statements that hit partitions attached via symlink may have slower runtimes
if their volumes have a slower disk.

:::

#### Properties using symbolic links

Partitions attached via the symbolic link approach are considered **read only**
by these operations:

    - _DROP/DETACH PARTITION_: Once the partition folders are unlinked, the symbolic links are removed, but not the content remains. Detaching a partition that was attached via symlink does not create a copy `<partition_name>.detached`; the operation is equal to DROP PARTITION.

Partitions attached via symlink are **read write** for these operations:

    - _INSERT_, _ADD/DROP INDEX_, _ADD/DROP/RENAME COLUMN_: Data in the partition folders

>       are modified in place as a result of the operation. The name of the symbolic link will change as a result.

Reviewer: >The name of the symbolic link will change as a result.
Isn't this mean that further operations won't be possible, because the name is changed?

>     - _UPDATE_: the partition folder's data files are modified in place.

Reviewer: What is the difference between this and L64?

Author: none, I will merge these two items.

## Description

Before executing `ATTACH PARTITION`, the partition folders to be attached must
be made available to QuestDB using one of the following methods:

- Copying the partition folders manually
- Using a [symbolic link](https://en.wikipedia.org/wiki/Symbolic_link)

This section describes the details of each method.

### Manual copy

Partition folders can be manually moved from where they are stored into the table folder in
`db`. To make the partitions available for the attach operation, the files
need to be renamed `<partition_name>.attachable`.

For example, in a table partitioned by year, given a partition folder named
`2020.detached`, rename it as `2020.attachable`, and move it to the table folder.

### Symbolic links

[Symbolic links](https://en.wikipedia.org/wiki/Symbolic_link) refer to partition folders that
do exist potentially in a different volume referred to as cold storage. To make cold detached partition 
folders available for attaching, create a symlink of name `<partition_name>.attachable` from within the table's folder
with target the path to the cold detached partition folder.

:::info

> 
> SQL statements that hit partitions attached via symbolic links may have slower runtimes
> if their volumes have a slower disk.
> 
> In windows, symlinks require admin privileges, and thus it is not recommended its use.

Reviewer: ```suggestion

- SQL statements that hit partitions attached via symbolic links may have slower runtimes
if their volumes have a slower disk.

- In Windows, symbolic links require admin privileges, and thus this method is not recommended.
```

```

You can now move those `<partition_name.detached>` folders to a different path, potentially a different volume:

```shell
mv ~/QUEST/db/tab/*.detached ~/QUEST/cool/tab

questdb@NYC:~/QUEST/db/tab$ ls -l
total 200
drwxrwxr-x  6 questdb  staff    192 Nov  8 14:58 2028.5
-rw-r--r--  1 questdb  staff  16384 Nov  8 14:58 _cv
-rw-r--r--  1 questdb  staff  16384 Nov  8 14:58 _meta
-rw-r--r--  1 questdb  staff  16384 Nov  8 14:58 _todo_
-rw-r--r--  1 questdb  staff  16384 Nov  8 14:58 _txn
-rw-r--r--  1 questdb  staff  32792 Nov  8 14:58 _txn_scoreboard
drwxrwxr-x  6 questdb  staff    192 Nov  8 14:58 seq
```

When you want them back, simply create a symlink:

```shell

questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2022.detached 2022.attachable
questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2023.detached 2023.attachable
questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2024.detached 2024.attachable
questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2025.detached 2025.attachable
questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2026.detached 2026.attachable
questdb@NYC:~/QUEST/db/tab$ ln -s ../../cool/tab/2027.detached 2027.attachable

> questdb@NYC:~/QUEST/db/tab$ ls -l

Author: listing the folder, I will amend.

total 200

> lrwxr-xr-x  1 marregui  staff     28 Nov  8 15:13 2022.attachable -> ../../cool/tab/2022.detached

Reviewer: Should we try to remove your name and make this more generic?

- The SQL reference to the partitions does not include the suffix `.attachable`.
- The `WHERE` clause is not supported when attaching partitions.

> - The very last partition cannot be detached, however it can be dropped.

Reviewer: ```suggestion
- The latest partition cannot be detached. However, it can be irreversibly deleted using [DROP TABLE](/docs/reference/sql/drop/).
```
